java/memory/memory-core/src/main/java/org/apache/arrow/util/Collections2.java

```diff
@@ -46,23 +45,55 @@ private Collections2() {}
    * Converts the iterable into a new {@link List}.
    */
   public static <T> List<T> toList(Iterable<T> iterable) {
-    return StreamSupport.stream(iterable.spliterator(), false).collect(Collectors.toList());
+    if (iterable instanceof Collection<?>) {
+      // If iterable is a collection, take advantage of it for a more efficient copy
+      return new ArrayList<T>((Collection<T>) iterable);
+    }
+    return toList(iterable.iterator());
   }
 
+  /**
+   * Converts the iterable into a new immutable {@link List}.
+   */
+  public static <T> List<T> toImmutableList(Iterable<T> iterable) {
+    return Collections.unmodifiableList(toList(iterable));
```

Comment:
Can we specialize this so it avoids a copy if the existing set is unmodifiable with a safe backing. Same with the unmodifiable map below. Seems like you could then keep the defensive copies you've removed below in Schema and FieldType

Reply:
Sad. Any chance we can source something small that already does that (in a separate jira)? Otherwise I expect we'll probably need another patch here in the future since we're still probably doing excessive copies.

Reply:
Not to knowledge